src/openInBrowser.js

```diff
@@ -0,0 +1,43 @@
+const opn = require('opn');
+const execSync = require('child_process').execSync;
+
+const openInBrowser = (url, browser) => {
+  const OSX_CHROME = 'google chrome';
+
+  const isBrowserString = typeof browser === 'string';
+
+  const shouldTryOpenChromeWithAppleScript =
+    process.platform === 'darwin' &&
+    (!isBrowserString || browser === OSX_CHROME);
+
+  if (shouldTryOpenChromeWithAppleScript) {
+    try {
+      // check if chrome is running
+      execSync('ps cax | grep "Google Chrome"');
+      const command = `osascript -l JavaScript reuseChromeTabInMacOS.js ${encodeURI(
```

Comment:
This is still using AppleScript (even though it’s in JavaScript), so this won’t work on other platforms etc.

I would much rather have us implement a nice and neat cross-platform solution that works for everyone. Could we maybe try to implement this using some kind of “polling” on the HMR client that starts when the connection get’s closed? That way we can just check for those incoming connections to the HRR server to determine if a tab is already open.

To clarify: My idea is that when the connection to the HMR server get’s closed, the HMR client should keep trying to reconnect or at least sending some kind or request. Then, we can just prevent the “open” feature from going off whenever we see these incoming connections.

Reply:
sorry, maybe i forgot to mention this is just a work around specifically for macOS on chrome.
i am not sure how to implement your suggestion and also if focusing on the correct browser window/tab has a pure js cross platform solution.
i liked this feature from create-react-app and was trying to replicate it.

i can look into into it later though.

Reply:
@davidnagli I searched more about it. 
- Just refreshing disconnected connection won't be enough, the browser window or the tab maybe in background.
Bringing the window and tab to front is important otherwise user might not know that the tab has been refreshed and will eventually open a new tab, thus leading to bad UX/DX.
- Supporting other browser and platform
  - MacOS
    - Applescript or mac os automation by javascript (**not using applescript**) can help bringing tab to focus.
    - Chrome is quite scriptable, thus we can provide this feature.
    - Firefox doesn't expose tabs api.
    - Safari does expose tab api but no way to bring it to front (activate). or directly reloading it
  - Windows
    - Can't automate chrome or firefox with vbscript
    - IE automation is possible, but i didn't look much into it.
  - Linux
    - Don't have access to a linux machine, maybe someone can look into.

I completely agree with you that the current solution is too specific (chrome on mac), but isn't it better to give atleast a few dev a good DX  than having no solution at all. 
If and when someone figure out a better way, this implementation can get replaced.

Reply:
How about we split these features into two seperate PRs.

It looks like the `--open [browser]` feature looks ready, and I’m a huge fan of it so I’d love to see that get merged as soon as possible. But I think we should discuss the other feature a bit further to work out the kinks.

Do you mind moving the “tab reuse feature” into a separate PR so we can discuss it separately from the “browser selection feature"?

Reply:
Sounds good. i'll make a PR soon